src/firebase_functions/options.py

```diff
@@ -0,0 +1,316 @@
+    The maximum value for concurrency is 1,000.
+    """
+
+    cpu: int | _typing.Literal["gcf_gen1"] | _Sentinel = "gcf_gen1"
+    """
+    Fractional number of CPUs to allocate to a function.
+    Defaults to 1 for functions with <= 2GB RAM and increases for larger memory sizes.
+    This is different from the defaults when using the gcloud utility and is different from
+    the fixed amount assigned in Google Cloud Functions generation 1.
+    To revert to the CPU amounts used in gcloud or in Cloud Functions generation 1, set this
+    to the value "gcf_gen1"
+    """
+
+    vpc_connector: str | _Sentinel | None = None
+    """
+    Connect cloud function to specified VPC connector.
+    A value of USE_DEFAULT removes the VPC connector.
+    """
+
+    vpc_connector_egress_settings: VpcEgressSettings | _Sentinel | None = None
+    """
+    Egress settings for VPC connector.
+    A value of USE_DEFAULT turns off VPC connector egress settings.
+    """
+
+    service_account: str | _Sentinel | None = None
+    """
+    Specific service account for the function to run as.
+    A value of USE_DEFAULT restores the default service account.
+    """
+
+    ingress: IngressSettings | _Sentinel | None = None
+    """
+    Ingress settings which control where this function can be called from.
+    A value of USE_DEFAULT turns off ingress settings.
+    """
+
+    labels: dict[str, str] | None = None
+    """
+    User labels to set on the function.
+    """
+
+    secrets: list[str] | list[SecretParam] | _Sentinel | None = None
+    """
+    Secrets to bind to a function.
+    """
+
+
+_GLOBAL_OPTIONS = _CommonOptions()
+"""The current default options for all functions. Internal use only."""
+
+
+@_dataclasses.dataclass(frozen=True)
+class _ProviderOptions:
+    """
+    Base class for provider specific options.
+    """
+
+    # We don't inherit _CommonOptions since some options can be overloaded
+    # with new types which will break inheritance.
+    options: _CommonOptions
+    """
+    Base options.
+    """
+
+    def asdict_with_global_options(self) -> dict:
+        """
+        Returns the provider options merged with globally defined options.
+        """
+        provider_options = _dataclasses.asdict(self)
+        # Copy base options to root of provider options.
+        provider_options = {**provider_options["options"], **provider_options}
+        del provider_options["options"]
+        global_options = _dataclasses.asdict(_GLOBAL_OPTIONS)
+        merged_options: dict = {**global_options, **provider_options}
+        # None values in the providers options should fallback to
+        # global options.
+        for key in provider_options:
+            if provider_options[key] is None and key in global_options:
+                merged_options[key] = global_options[key]
+        # None values are automatically stripped out in ManifestEndpoint generation.
+
+        # TODO how should _Sentinel values be translated to spec?
```

Comment:
Tracking separately to address. I will probably address this inside the manifest generation since I think Sentinel values only need to be dropped from resulting manifest output since CF3 backend handles defaults?

Reply:
_Sentinel should show up in the YAML spec as a `null`.